[PyTorch] Add MaybeOwned::operator*() &&

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55351 [PyTorch] Fix const correctness for resize native functions
* #55685 [PyTorch] Avoid double indirection in MaybeOwned's borrowed state
* #55684 [PyTorch] Change MaybeOwned tests to use intrusive_ptr and Tensor
* #55553 [PyTorch] Mark borrowed case as C10_LIKELY in MaybeOwned
* #55419 [PyTorch] Allow copy operations on MaybeOwned
* #55316 [PyTorch] Add & use inferExpandGeometry_dimvector
* #55247 [PyTorch][easy] Use sizes()[x] instead of size(x) in addr
* #55246 [PyTorch] Update expand_size API to match expand_inplace
* #55245 [PyTorch] Update expand_outplace API to match expand_inplace
* **#55244 [PyTorch] Add MaybeOwned::operator*() &&**
* #55180 [PyTorch] Use infer_size_dimvector in ExpandUtils
* #55065 [PyTorch] New expand_inplace API with MaybeOwned<Tensor> and no unary tuples

Add the ability to move from the underlying object in a `MaybeOwned`.

FWIW, `MaybeOwned` is new territory for me personally and this move-and-dereference operation is even more so, but I think it makes sense and the tests pass.

Differential Revision: [D27522809](https://our.internmc.facebook.com/intern/diff/D27522809/)

swolchok committed:
[PyTorch] Add MaybeOwned::operator*() &&

Add the ability to move from the underlying object in a `MaybeOwned`.

FWIW, `MaybeOwned` is new territory for me personally and this move-and-dereference operation is even more so, but I think it makes sense and the tests pass.

Differential Revision: [D27522809](https://our.internmc.facebook.com/intern/diff/D27522809/)

[ghstack-poisoned]

swolchok committed:
rebase, fix typo in D27522811 on "[PyTorch] Add MaybeOwned::operator*() &&"


Add the ability to move from the underlying object in a `MaybeOwned`.

FWIW, `MaybeOwned` is new territory for me personally and this move-and-dereference operation is even more so, but I think it makes sense and the tests pass.

Differential Revision: [D27522809](https://our.internmc.facebook.com/intern/diff/D27522809/)

[ghstack-poisoned]

swolchok commented:
ping

swolchok committed:
see if OSS doctest failures about non-ASCII character go away on "[PyTorch] Add MaybeOwned::operator*() &&"


Add the ability to move from the underlying object in a `MaybeOwned`.

FWIW, `MaybeOwned` is new territory for me personally and this move-and-dereference operation is even more so, but I think it makes sense and the tests pass.

Differential Revision: [D27522809](https://our.internmc.facebook.com/intern/diff/D27522809/)

[ghstack-poisoned]

swolchok commented:
ping -- it's been at least a day with no review

bhosmer commented on c10/util/MaybeOwned.h:
```diff
@@ -103,6 +103,19 @@ class MaybeOwned final {
+  // T.
+  T operator*() && {
+    if (isBorrowed_) {
+      TORCH_INTERNAL_ASSERT_DEBUG_ONLY(borrow_ != nullptr);
```
No action needed, but curious the motivation behind checking this on derefs rather than on construction/assignment

swolchok replied:
Default construction creates a null borrowed MaybeOwned to simplify initialization inside an if statement.

swolchok committed:
Update on "[PyTorch] Add MaybeOwned::operator*() &&"


Add the ability to move from the underlying object in a `MaybeOwned`.

FWIW, `MaybeOwned` is new territory for me personally and this move-and-dereference operation is even more so, but I think it makes sense and the tests pass.

Differential Revision: [D27522809](https://our.internmc.facebook.com/intern/diff/D27522809/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add MaybeOwned::operator*() &&"


Add the ability to move from the underlying object in a `MaybeOwned`.

FWIW, `MaybeOwned` is new territory for me personally and this move-and-dereference operation is even more so, but I think it makes sense and the tests pass.

Differential Revision: [D27522809](https://our.internmc.facebook.com/intern/diff/D27522809/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add MaybeOwned::operator*() &&"


Add the ability to move from the underlying object in a `MaybeOwned`.

FWIW, `MaybeOwned` is new territory for me personally and this move-and-dereference operation is even more so, but I think it makes sense and the tests pass.

Differential Revision: [D27522809](https://our.internmc.facebook.com/intern/diff/D27522809/)

[ghstack-poisoned]
